File: src/xai/sdk/python/xai_sdk/clients/wallet_client.py
"""
Wallet Client for XAI SDK

Handles all wallet-related operations.
"""

from typing import Optional, Dict, Any
from datetime import datetime

from ..http_client import HTTPClient
from ..models import Wallet, Balance, WalletType
from ..exceptions import WalletError, ValidationError


class WalletClient:
    """Client for wallet operations."""

    def __init__(self, http_client: HTTPClient) -> None:
        """
        Initialize Wallet Client.

        Args:
            http_client: HTTP client instance
        """
        self.http_client = http_client

    def create(
        self,
        wallet_type: WalletType = WalletType.STANDARD,
        name: Optional[str] = None,
    ) -> Wallet:
        """
        Create a new wallet.

        Args:
            wallet_type: Type of wallet to create
            name: Optional wallet name

        Returns:
            Created wallet

        Raises:
            WalletError: If wallet creation fails
        """
        try:
            payload = {
                "wallet_type": wallet_type.value,
            }
            if name:
                payload["name"] = name

            response = self.http_client.post("/wallet/create", data=payload)

            return Wallet(
                address=response["address"],
                public_key=response["public_key"],
                created_at=datetime.fromisoformat(response["created_at"]),
                wallet_type=WalletType(response.get("wallet_type", "standard")),
                private_key=response.get("private_key"),
            )
        except Exception as e:
            raise WalletError(f"Failed to create wallet: {str(e)}")

    def get(self, address: str) -> Wallet:
        """
        Get wallet information.

        Args:
            address: Wallet address

        Returns:
            Wallet details

        Raises:
            WalletError: If wallet retrieval fails
        """
        if not address:
            raise ValidationError("Address is required")

        try:
            response = self.http_client.get(f"/wallet/{address}")
            return Wallet(
                address=response["address"],
                public_key=response["public_key"],
                created_at=datetime.fromisoformat(response["created_at"]),
                wallet_type=WalletType(response.get("wallet_type", "standard")),
                nonce=response.get("nonce", 0),
            )
        except Exception as e:
            raise WalletError(f"Failed to get wallet: {str(e)}")

    def get_balance(self, address: str) -> Balance:
        """
        Get wallet balance.

        Args:
            address: Wallet address

        Returns:
            Balance information

        Raises:
            WalletError: If balance retrieval fails
        """
        if not address:
            raise ValidationError("Address is required")

        try:
            response = self.http_client.get(f"/wallet/{address}/balance")
            return Balance(
                address=response["address"],
                balance=response["balance"],
                locked_balance=response.get("locked_balance", "0"),
                available_balance=response.get("available_balance", response["balance"]),
                nonce=response.get("nonce", 0),
            )
        except Exception as e:
            raise WalletError(f"Failed to get balance: {str(e)}")

    def get_transactions(
        self,
        address: str,
        limit: int = 50,
        offset: int = 0,
    ) -> Dict[str, Any]:
        """
        Get wallet transaction history.

        Args:
            address: Wallet address
            limit: Number of transactions to retrieve
            offset: Offset for pagination

        Returns:
            Transaction history with metadata

        Raises:
            WalletError: If transaction retrieval fails
        """
        if not address:
            raise ValidationError("Address is required")

        if limit > 100:
            limit = 100

        try:
            response = self.http_client.get(
                f"/wallet/{address}/transactions",
                params={"limit": limit, "offset": offset},
            )
            return {
                "transactions": response.get("transactions", []),
                "total": response.get("total", 0),
                "limit": response.get("limit", limit),
                "offset": response.get("offset", offset),
            }
        except Exception as e:
            raise WalletError(f"Failed to get transactions: {str(e)}")

    def create_embedded(
        self,
        app_id: str,
        user_id: str,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        """
        Create an embedded wallet.

        Args:
            app_id: Application ID
            user_id: User ID
            metadata: Optional metadata

        Returns:
            Embedded wallet information

        Raises:
            WalletError: If embedded wallet creation fails
        """
        if not app_id or not user_id:
            raise ValidationError("app_id and user_id are required")

        try:
            payload = {
                "app_id": app_id,
                "user_id": user_id,
            }
            if metadata:
                payload["metadata"] = metadata

            return self.http_client.post("/wallet/embedded/create", data=payload)
        except Exception as e:
            raise WalletError(f"Failed to create embedded wallet: {str(e)}")

    def login_embedded(self, wallet_id: str, password: str) -> Dict[str, Any]:
        """
        Login to an embedded wallet.

        Args:
            wallet_id: Embedded wallet ID
            password: Wallet password

        Returns:
            Session information

        Raises:
            WalletError: If login fails
        """
        if not wallet_id or not password:
            raise ValidationError("wallet_id and password are required")

        try:
            payload = {
                "wallet_id": wallet_id,
                "password": password,
            }
            return self.http_client.post("/wallet/embedded/login", data=payload)
        except Exception as e:
            raise WalletError(f"Failed to login to embedded wallet: {str(e)}")

File: src/xai/sdk/typescript/src/clients/wallet-client.ts
/**
 * Wallet Client for XAI SDK
 *
 * Handles all wallet-related operations including creation, balance queries,
 * and transaction history.
 */

import { HTTPClient } from '../utils/http-client';
import {
  Wallet,
  Balance,
  WalletType,
  CreateWalletParams,
  CreateEmbeddedWalletParams,
  EmbeddedWalletLoginParams,
  TransactionHistoryParams,
  PaginatedResponse,
  Transaction,
} from '../types';
import { WalletError, ValidationError } from '../errors';

/**
 * Client for wallet operations
 */
export class WalletClient {
  constructor(private httpClient: HTTPClient) {}

  /**
   * Create a new wallet
   *
   * @param params - Wallet creation parameters
   * @returns Created wallet with address and keys
   *
   * @example
   * ```typescript
   * const wallet = await client.wallet.create({
   *   walletType: WalletType.STANDARD,
   *   name: 'My Wallet'
   * });
   * console.log('New wallet address:', wallet.address);
   * ```
   */
  async create(params: CreateWalletParams = {}): Promise<Wallet> {
    try {
      const payload: Record<string, unknown> = {};

      if (params.walletType) {
        payload.wallet_type = params.walletType;
      }
      if (params.name) {
        payload.name = params.name;
      }

      const response = await this.httpClient.post<{
        address: string;
        public_key: string;
        created_at: string;
        wallet_type?: string;
        private_key?: string;
      }>('/wallet/create', payload);

      return {
        address: response.address,
        publicKey: response.public_key,
        createdAt: response.created_at,
        walletType: response.wallet_type as WalletType,
        privateKey: response.private_key,
      };
    } catch (error) {
      if (error instanceof WalletError) {
        throw error;
      }
      throw new WalletError(`Failed to create wallet: ${error}`);
    }
  }

  /**
   * Get wallet information
   *
   * @param address - Wallet address
   * @returns Wallet details
   *
   * @example
   * ```typescript
   * const wallet = await client.wallet.get('0x1234...');
   * console.log('Wallet nonce:', wallet.nonce);
   * ```
   */
  async get(address: string): Promise<Wallet> {
    if (!address) {
      throw new ValidationError('Address is required');
    }

    try {
      const response = await this.httpClient.get<{
        address: string;
        public_key: string;
        created_at: string;
        wallet_type?: string;
        nonce?: number;
      }>(`/wallet/${address}`);

      return {
        address: response.address,
        publicKey: response.public_key,
        createdAt: response.created_at,
        walletType: response.wallet_type as WalletType,
        nonce: response.nonce || 0,
      };
    } catch (error) {
      if (error instanceof WalletError) {
        throw error;
      }
      throw new WalletError(`Failed to get wallet: ${error}`);
    }
  }

  /**
   * Get wallet balance
   *
   * @param address - Wallet address
   * @returns Balance information
   *
   * @example
   * ```typescript
   * const balance = await client.wallet.getBalance('0x1234...');
   * console.log('Balance:', balance.balance);
   * console.log('Available:', balance.availableBalance);
   * ```
   */
  async getBalance(address: string): Promise<Balance> {
    if (!address) {
      throw new ValidationError('Address is required');
    }

    try {
      const response = await this.httpClient.get<{
        address: string;
        balance: string;
        locked_balance?: string;
        available_balance?: string;
        nonce?: number;
      }>(`/wallet/${address}/balance`);

      return {
        address: response.address,
        balance: response.balance,
        lockedBalance: response.locked_balance || '0',
        availableBalance: response.available_balance || response.balance,
        nonce: response.nonce || 0,
      };
    } catch (error) {
      if (error instanceof WalletError) {
        throw error;
      }
      throw new WalletError(`Failed to get balance: ${error}`);
    }
  }

  /**
   * Get wallet transaction history
   *
   * @param params - Query parameters with address, limit, and offset
   * @returns Paginated transaction history
   *
   * @example
   * ```typescript
   * const history = await client.wallet.getTransactions({
   *   address: '0x1234...',
   *   limit: 20,
   *   offset: 0
   * });
   * console.log('Transactions:', history.data);
   * console.log('Total:', history.total);
   * ```
   */
  async getTransactions(params: TransactionHistoryParams): Promise<PaginatedResponse<Transaction>> {
    if (!params.address) {
      throw new ValidationError('Address is required');
    }

    const limit = Math.min(params.limit || 50, 100);
    const offset = params.offset || 0;

    try {
      const response = await this.httpClient.get<{
        transactions: Transaction[];
        total: number;
        limit: number;
        offset: number;
      }>(`/wallet/${params.address}/transactions`, {
        limit,
        offset,
      });

      return {
        data: response.transactions || [],
        total: response.total || 0,
        limit: response.limit || limit,
        offset: response.offset || offset,
      };
    } catch (error) {
      if (error instanceof WalletError) {
        throw error;
      }
      throw new WalletError(`Failed to get transactions: ${error}`);
    }
  }

  /**
   * Create an embedded wallet
   *
   * @param params - Embedded wallet creation parameters
   * @returns Embedded wallet information
   *
   * @example
   * ```typescript
   * const embeddedWallet = await client.wallet.createEmbedded({
   *   appId: 'my-app',
   *   userId: 'user-123',
   *   metadata: { email: 'user@example.com' }
   * });
   * ```
   */
  async createEmbedded(params: CreateEmbeddedWalletParams): Promise<Record<string, unknown>> {
    if (!params.appId || !params.userId) {
      throw new ValidationError('appId and userId are required');
    }

    try {
      const payload: Record<string, unknown> = {
        app_id: params.appId,
        user_id: params.userId,
      };

      if (params.metadata) {
        payload.metadata = params.metadata;
      }

      return await this.httpClient.post('/wallet/embedded/create', payload);
    } catch (error) {
      if (error instanceof WalletError) {
        throw error;
      }
      throw new WalletError(`Failed to create embedded wallet: ${error}`);
    }
  }

  /**
   * Login to an embedded wallet
   *
   * @param params - Login parameters with wallet ID and password
   * @returns Session information
   *
   * @example
   * ```typescript
   * const session = await client.wallet.loginEmbedded({
   *   walletId: 'wallet-123',
   *   password: 'secure-password'
   * });
   * ```
   */
  async loginEmbedded(params: EmbeddedWalletLoginParams): Promise<Record<string, unknown>> {
    if (!params.walletId || !params.password) {
      throw new ValidationError('walletId and password are required');
    }

    try {
      const payload = {
        wallet_id: params.walletId,
        password: params.password,
      };

      return await this.httpClient.post('/wallet/embedded/login', payload);
    } catch (error) {
      if (error instanceof WalletError) {
        throw error;
      }
      throw new WalletError(`Failed to login to embedded wallet: ${error}`);
    }
  }
}
